src/sparseml/optim/helpers.py

```diff
@@ -512,3 +518,118 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
             return float(val)
         except Exception:
             return val
+
+
+def _extract_metadata_from_recipe(container):
+    metadata = {}
+    if RECIPE_METADATA_KEY in container.keys():
+        metadata = container[RECIPE_METADATA_KEY]
+    return metadata
+
+
+def _extract_metadata_from_staged_recipe(container):
+    metadata = {}
+    stage_names = _get_recipe_stage_names(container)
+    for stage_name in stage_names:
+        if RECIPE_METADATA_KEY in container[stage_name].keys():
+            metadata[stage_name] = container[stage_name][RECIPE_METADATA_KEY]
+    if metadata and (len(metadata) != len(stage_names)):
+        raise ValueError(
+            "It seems that some stages in your checkpoint recipe"
+            "contain metadata and some don't."
+            "This is not allowed: either all the "
+            f"stages contain {RECIPE_METADATA_KEY} key or none does."
+        )
+
+    return metadata
+
+
+def _get_recipe_stage_names(container):
+    # Extracts valid stage names from a container.
+    # Valid stage name (key) is the one which corresponds to a value, which is
+    # a dictionary where at least one of the keys contains a string 'modifiers'.
+
+    stage_names = [
+        stage_name
+        for stage_name, stage_dict in container.items()
+        if isinstance(stage_dict, dict)
+        and any([key for key in stage_dict.keys() if "modifiers" in key])
+    ]
+    return stage_names
+
+
+def _check_warn_dict_difference(original_dict, new_dict):
+    if original_dict != new_dict:
+        warnings.warn(
+            f"Attempting to overwrite old metadata: {original_dict} "
+            f"with new metadata: {new_dict}. "
+            "It is prohibitive to alter the metadata of the recipe. "
+            "The new metadata will be omitted and discarded."
```

Comment:
```suggestion
            "This may lead to different results than the original run of the recipe. "
            "The previous metadata will be omitted and discarded. Ignore if a "
            "change in metadata is expected"
```
The old metadata... correct?

Reply:
Wait. This is something which we are not in sync still.

You suggest that if we have `previous_metadata` and pass `metadata` and they are not equal, we should raise a warning, yet still ALLOW to completely overwrite the `previous_metadata` with the new `metadata` right? This means wiping `previous_metadata` completely and replacing it with `metadata`. i.e. we trust the user that they know what they are doing if they decide to pass `metadata` along with a recipe that already contains metadata. 
E.g.
`previous metadata = {x:2, y:3, z:4}` and `metadata = {u:5}` --> result will be `{u:5}` right?

Reply:
At the end of the day, it is really about how much we trust the user. I will change the functionality of my code to behave as described above.

Reply:
yes, this is what we determined. Once we start a new training run with a recipe, the previous metadata from that recipe is now irrelevant.  In your example, if x,y, and z no longer exist or have had their values changed in the new run, why should they be tracked by the new run just because they were set previously?  More concretely, say a checkpoint run was done at batch size 16 and we launch a new run that for some reason does not track batch size in metadata - it would be wrong to assume that the new run is using batch size 16 and thus should not be considered in the new metadata.

Please make sure that the logic is consistent with this - the previous metadata should always be overwritten if new metadata is given.

Reply:
Got it. This is the current behavior now, I confirm. Will double check tho (or quickly write some more tests if necessary)